### PR TITLE
Add function-specific keyword-arguments to metric functions.

### PR DESCRIPTION
**Pull request by dkrako**
This PR fixes the main issue from #82 

This is done, as the same arguments passed via the unspecific `kwargs`-argument are
then further passed for example to both `perturb_func()` and `model.predict()`.

This commit therefore introduces `normalise_func_kwargs`, `model_predict_kwargs` and
`perturb_func_kwargs`. For the current version the actual behaviour will not be
changed, so passing unspecific `kwargs` is still working as before, i.e.
they are being passed to `normalise_func` and `perturb_func`.
This should be deprecated and eventually removed in future versions.

For this feature, the `base.Metric` class was extended to contain all the
general attributes and preprocessing which was present in each specific
metric. Most of this is now handled by the `prepare()` method of the
base-class. There is also the `base.PerturbationMetric` class added, which
handles the perturbation attributes. This way maintainability and
readability is much improved, as fixing issues can be potentially solved
in the base-class only, without needing to go through all of the
metrics.

Each metric now has to implement a `process_instance()` method, where a
single instance of the input batch is processed and the score is simply
returned. The scores will then be written in the base-class into the `last_results`
attribute. Addtionally, each metric can implement `preprocess()` and
`postprocess()` methods. If the structure of the base-class doesn't fit
this design, it can still be implemented the old way, like seen in the
metric `ModelParameterRandomisation`, which nevertheless still uses the handy `prepare()`
method from the base class.

### CODE REVIEW

**Comment by annahedstroem** [comment]
Fantastic @dkrako. More in-depth review to come.

In the meanwhile, pasting ome comments for later review:
- Asserts, deprecation warnings
- Check that progress bar works
- How about device?
- Dev safety mechanisms:
    - Kwargs append… if key is overwritten: write a print (could add later, now add a print)
    - Check if defaults are used?
- Naming: call it custom_preprocess

File: quantus/evaluation.py [hunk 1]
```diff
@@ -72,6 +73,7 @@ def evaluate(
                         y_batch=y_batch,
                         a_batch=a_batch,
                         s_batch=s_batch,
+                        # TODO: what does the method parameter do?
```
Comment (leanderweber) [inline quantus/evaluation.py:76]: I believe this is a kwarg for the explain_function. The quantus wrappers around tf_explain and captum take a "method" kwarg. This does not fit with zennit though.

I would suggest we change this a bit. xai_methods should be a dict mapping {explain_method_name: explain_method_kwargs}. The methodname can be used for results, and the method_kwargs to compute explanations. This would fit very well with this PR, and make this xai evaluation a lot more customizable.

File: quantus/helpers/normalise_func.py [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
 
-def normalise_by_max(a: np.ndarray) -> np.ndarray:
-    """ "Normalize attributions by the maximum absolute value of the explanation."""
+def normalise_batch(
+        arr: np.ndarray,
+        normalise_func: Callable,
+        **normalise_func_kwargs,
+) -> None:
+    """Normalise complete batch by given normalisation function."""
+    # TODO: use something like np.apply_along_axis instead of for-loop
```
Comment (leanderweber) [inline quantus/helpers/normalise_func.py:12]: I would suggest something like this:

1. Add an axes argument to all normalisation functions, to specify axes over which the normalisation is performed.
2. Use this argument to exclude the batch axis during normalisation

File: quantus/helpers/pytorch_model.py [hunk 3]
```diff
@@ -13,37 +13,59 @@
...
 
-        softmax = kwargs.get("softmax", False)
-        device = kwargs.get("device", None)
-        grad = kwargs.get("grad", False)
         grad_context = torch.no_grad() if not grad else suppress()
 
         with grad_context:
-            pred = self.model(torch.Tensor(x).to(device))
-            if softmax:
+            pred = self.model(torch.Tensor(x).to(self.device), **predict_kwargs)
```
[the remark this answers is not in the document]
Comment (leanderweber) [reply quantus/helpers/pytorch_model.py:43]: Looked it up, and apparently it does. nice!

File: quantus/metrics/axiomatic_metrics.py [hunk 4]
```diff
@@ -521,133 +563,123 @@ def __init__(self, *args, **kwargs):
...
+        a_shifted = self.explain_func(
+            model=model.get_model(), inputs=x_shifted, targets=y,
+            **self.explain_func_kwargs
+        )
 
-        # Use tqdm progressbar if not disabled.
-        if not self.display_progressbar:
-            iterator = zip(x_batch_s, y_batch, a_batch)
+        # Check if explanation of shifted input is similar to original.
+        if (a.flatten() != a_shifted.flatten()).all():
```
Comment (leanderweber) [inline quantus/metrics/axiomatic_metrics.py:668]: Sidenote unrelated to this PR: This seems to be a really strict condition. Does it make sense to give some options here w.r.t. strictness (e.g., similarity_func(a, a_shifted, precision=something)). Also for similarly strict metrics.

File: quantus/metrics/base.py [hunk 5]
```diff
@@ -1,105 +1,331 @@
...
+            channel_first: Optional[bool] = None,
+            explain_func: Optional[Callable] = None,
+            explain_func_kwargs: Optional[Dict[str, Any]] = None,
+            softmax: bool = False,
+            device: Optional[str] = None,
+            model_predict_kwargs: Optional[Dict] = None,
+            **kwargs,
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, ModelInterface, Dict[str, Any]]:
+        '''
+        TODO: add documentation.
```
Comment (leanderweber) [inline quantus/metrics/base.py:220]: add documentation :)

File: quantus/helpers/normalise_func.py (hunk 2 above)
Comment (annahedstroem) [reply quantus/helpers/normalise_func.py:12]: +1

File: quantus/helpers/pytorch_model.py [hunk 6]
```diff
@@ -13,37 +13,59 @@
 class PyTorchModel(ModelInterface):
     """Interface for torch models."""
 
-    def __init__(self, model, channel_first):
-        super().__init__(model, channel_first)
+    def __init__(
```
Comment (annahedstroem) [inline quantus/helpers/pytorch_model.py:16]: Makes sense to put it all in __init__!

File: quantus/helpers/pytorch_model.py [hunk 7]
```diff
@@ -13,37 +13,59 @@
...
+        super().__init__(
+            model=model,
+            channel_first=channel_first,
+            softmax=softmax,
+            predict_kwargs=predict_kwargs,
+        )
+        self.device = device
 
-    def predict(self, x, **kwargs):
+    def predict(self, x, grad=False, **kwargs):
```
Comment (annahedstroem) [inline quantus/helpers/pytorch_model.py:31]: add typing

File: quantus/helpers/pytorch_model.py [hunk 8]
```diff
@@ -13,37 +13,59 @@
...
         """Predict on the given input."""
+
+        # Use kwargs of predict call if specified, but don't overwrite object attribute
+        predict_kwargs = {**self.predict_kwargs, **kwargs}
```
Comment (annahedstroem) [inline quantus/helpers/pytorch_model.py:35]: Will fix a solution for this.

**Review by annahedstroem — COMMENTED** [review]
Not finished with review.

File: quantus/metrics/base.py [hunk 9]
```diff
@@ -1,105 +1,331 @@
...
-    @attributes_check
-    def __init__(self, *args, **kwargs):
+    @asserts.attributes_check
+    def __init__(
+            self,
+            abs: bool = False,
+            normalise: bool = False,
+            normalise_func: Optional[Callable] = None,
+            normalise_func_kwargs: Optional[Dict[str, Any]] = None,
+            softmax: bool = False,
```
Comment (annahedstroem) [inline quantus/metrics/base.py:29]: Can't we have something called predict_kwargs? or model_kwargs?

File: quantus/metrics/base.py [hunk 10]
```diff
@@ -1,105 +1,331 @@
...
-        self.normalise_func = self.kwargs.get("normalise_func", normalise_by_negative)
-        self.default_plot_func = Callable
-        self.disable_warnings = self.kwargs.get("disable_warnings", False)
-        self.display_progressbar = self.kwargs.get("display_progressbar", False)
+        self.abs = abs
+        self.normalise = normalise
+        self.softmax = softmax
+
+        if normalise_func is None:
+            self.normalise_func = normalise_by_negative
```
Comment (annahedstroem) [inline quantus/metrics/base.py:57]: Why not add to signature? much clearer that `normalise_by_negative` is the default value then

File: quantus/metrics/base.py [hunk 11]
```diff
@@ -1,105 +1,331 @@
...
+        self.normalise_func = normalise_func
+        if normalise_func_kwargs is None:
+            normalise_func_kwargs = {}
+        # TODO: deprecate this kind of unspecific kwargs passing
+        # this code prioritizes normalise_func_kwargs items over kwargs items
+        normalise_func_kwargs = {**kwargs, **normalise_func_kwargs}
```
Comment (annahedstroem) [inline quantus/metrics/base.py:63]: Can't we add:

```python
if kwargs:
    warn_funcs.kwargs_warning()
```
where 

```python

def kwargs_warning:
    warnings.warn(f"Passing **kwargs in the metric initalisation or call will be deprecated in future releases. Make sure to specify your parameters in the respective parameter space such as 'normalise_func_kwargs', 'explain_func_kwargs', 'model_predict_kwargs', 'perturb_func_kwargs' etc ..." " DeprecationWarning)

```

File: quantus/metrics/base.py [hunk 12]
```diff
@@ -1,105 +1,331 @@
...
+            explain_func_kwargs=explain_func_kwargs,
+            softmax=softmax,
+            device=device,
+            model_predict_kwargs=model_predict_kwargs,
+            **kwargs,
+        )
+
+        # Create progress bar if desired.
+        iterator = tqdm(
+            enumerate(zip(x_batch, y_batch, a_batch, s_batch)),
```
Comment (annahedstroem) [inline quantus/metrics/base.py:163]: This can be problematic if we need to loop through the _batches more than once.... so we can't set len(x_batch) it also has to account for the # iterations, # runs etc, if that is apart of the metric.

File: quantus/metrics/base.py [hunk 13]
```diff
@@ -1,105 +1,331 @@
...
+            y_batch=y_batch,
+            a_batch=a_batch,
+            s_batch=s_batch,
+            channel_first=channel_first,
+            explain_func=explain_func,
+            explain_func_kwargs=explain_func_kwargs,
+            softmax=softmax,
+            device=device,
+            model_predict_kwargs=model_predict_kwargs,
+            **kwargs,
```
Comment (annahedstroem) [inline quantus/metrics/base.py:158]: Hmm.. what do we do with these?

File: quantus/metrics/base.py (hunk 12 above)
Comment (dkrako) [reply quantus/metrics/base.py:163]: I would concur, as you can still implement the multiple runs per batch in the `process_batch` method of each metric (this is implemented for example in the `PixelFlipping` and `RegionPerturbation` metrics).
so as long as the instances themselves are processed independently, this shouldn't be really problematic.
you just need to reorder the loop in your head.

for the progress bar, I don't think that it's necessary to have such a fine-granular resolution. having the number of batches as the denominator in the progress bar should be sufficient. at least for me, I don't see the need to have updates every second as long as I know how long a single iterations takes.
I would otherwise worry to introduce too much metric-specifics in the base-class, as it would then need something like a `nr_runs` parameter, but some metrics will then call this parameter differently and we have chaos again.

but the fix should still be easy if a more granular progressbar is really desired. I would propose to first test lower progressbar-granularity and then only change this if it shows to be really insufficient in practice.

File: quantus/metrics/base.py (hunk 11 above)
Comment (dkrako) [reply quantus/metrics/base.py:63]: Yes, this is a good idea to factor this out. I think we should also raise a warning instead of using print in `merge_kwargs`.

File: quantus/metrics/base.py (hunk 9 above)
Comment (dkrako) [reply quantus/metrics/base.py:29]: I put it in `__call__`, as I didn't think that it makes sense to have model kwargs before passing the model (which is passed just in `__call__`, not in `__init__`.

**Review by annahedstroem — CHANGES_REQUESTED** [review]

**Comment by dkrako** [comment]
As discussed I have left the `**kwargs` in each metric method but raise a `ValueError` with informative text if there are any unexpected keyword arguments passed to the base methods `__init__()` and `__call__()`.

See here: https://github.com/dkrako/Quantus/blob/30b656b2620ae7517f946d178d8636571000d12d/quantus/helpers/asserts.py#L253

**Review by dkrako — COMMENTED** [review]
I think the `set_params()` method now doesn't make sense anymore, at least with the unspecific kwargs variable.
https://github.com/dkrako/Quantus/blob/30b656b2620ae7517f946d178d8636571000d12d/quantus/metrics/base.py#L335

**Comment by dkrako** [comment]
I have updated `test_axiomatic_metrics.py` to propose the change we will need to apply to every metric test.
https://github.com/dkrako/Quantus/blob/feature/function-specific-kwargs/tests/metrics/test_axiomatic_metrics.py

I have nested the `params` dictionary by adding an `init` and `call` key, where the arguments for the respective methods will be stored.
If `a_batch_generate == True`, then `explain_func` and `explain_func_kwargs` will be read from the `call` params dict.

@annahedstroem  @leanderweber
Please leave me feedback here if this would be the appropriate change that we should apply for the remaining test cases.
